### DISCUSSION TITLE
[solved] Scripting: check Pin-States if on, off or "blinking"

### DISCUSSION THREAD

**Discussion by git5any** [arendst/Tasmota#15768] [Q&A] [answered]
Hi everyone,
I'm rather new to Tasmota scripting on a ESP32 and have some questions about it. Compiling etc works fine, I have scripting enabled, using it for SML-reading already. The used board is a WT32-ETH01, a ESP32 with Ethernet on board. Some GPIOs are not available, I testes the avaiable ones if I could use them in the script which works fine. Ethernet also works.
What I am trying to do is the following:
I have 4 LEDs on a equipment and get their states with a photodiode and a little circuit, so i have a nice digital high/low-signal if the LED is on/off. Now I need to check whether an LED is on, off or blinking. Blinking can be with two speeds (I name it slow and fast), meaning  1 sec or 2 sec between two "on"-states.
My approach is to check in the >F Section the GPIO-state, if there is a change to check, if its on or off and if a running timer is either approx. 0.5 sec or approx. 1 sec or more than 2 sec. The result is then 0(off), 1(on), 2(fastblink) or 3(slowblink) If the result changes it will be published via MQTT.
Some part of my script work, I can read the pin, determine the state, but the rest ist not working as expected. Publishing also works, for the time I only print to the console.

here is the script: (there are some variables defined not used at this moment)
```

>D
pnP=0 ;Value of Pin PowerLED
thP=50 ;timer High PowerLED
tlP=50 ;timer Low PowerLED
lP=0 ; lastValue PowerLED
rskP=4 ;Rueckstauklappe PowerLED
oP=4 ;oldValue_RueckstauklappePowerLED
stP=3 ;status PowerLED

ms=0 ;MQTTstate
mst="m_wait"  ;MQTTstateText
ws=0
wst="w_wait"
myTime="xx"


>B
->sensor53 r
->ButtonDebounce 130
->timezone 99
; gpio39 als Input (ohne Pullup)
spinm(39,0)
if pin[39]<0.1
  then
  rskP=0
  else
  rskP=1
endif
;rskP=0
oP=rskP
print pin39 = %pin[39]% WiFi: %wifis%
print (boot) rskp: %rskP%  oP: %oP%




>F
pnP=pin[39]
thP+=1
tlP+=1

if chg[pnP]>0 ;check for change on GPIO
	then
	print change on GPIO
	if pnP<0.1 ;(LED off)
		then
		print Pin = %pnP%
		if thP>25
			then
			lP=1
			rskP=1
		else
		if thP>7
		and thP<13
			then
			lP=1
			rskP=3
		else
		if thP>2
		and thP<8
			then
			lP=1
			rskP=2
		endif
		thP=0 ; reset high counter
	else
	if pnP>0.1 ;(LED on)
		then
		print Pin = %pnP%
		if tlP>25
			then
			lP=0
			rskP=0
		else
		if tlP>7
		and tlP<13
			then
			lP=0
			rskP=3
		else
		if tlP>2
		and tlP<8
			then
			lP=0
			rskP=2
		endif
		tlP=0 ; reset low counter
	endif
	if rskP!=oP
		then
		;=>Publish tele/%topic%/SENSOR {"RSKl":{"P_LED":%rskP%}}
		print new output(low) rskP: %rskP%, oP: %oP%
		oP=rskP
	endif
endif

if tlP > 500
then
tlP=50
endif

if thP > 500
then
thP=50
endif


>S
; prints the counter-vars and the result (for testing)
print  thP: %thP%  tlP: %tlP% rskP: %rskP%  oP: %oP%


```
script starts without errors:

12:54:48.450 Script: nv=9, tv=3, vns=48, vmem=232, smem=8192
`
The change on the GPIO is recognized, but only a change to low is then printed. Timer-variables are never reset. I think, the if-then-else-endif are correct, but I might be wrong. There are many ways for inputs: Switch, button, input or like I did in the script.  Maybe my approach is completely noob, so if you have suggestions please feel free. I also found mpt(x) in the docs, maybe that could be used? I don't understand the usage. Finally I need that for 4 LEDs, could be I have to compile with more "space" for the script later on.

Thank you very much for any help!

**Comment by gemu2015** [comment]
I probably would choose another approach 
Use tasmota counters to count the led state transitions. They are IRQ driven and will not miss a change.
You may even set the counters into frequency mode. Don’t know if this works also with such low values below 1 hertz but this might be the easiest way.
Or if you use sml on the same time you may also use sml counter channels
And then in regular intervals check the counter state

**Reply by git5any** [reply]
> I probably would choose another approach Use tasmota counters to count the led state transitions. They are IRQ driven and will not miss a change. You may even set the counters into frequency mode. Don’t know if this works also with such low values below 1 hertz but this might be the easiest way. Or if you use sml on the same time you may also use sml counter channels And then in regular intervals check the counter state

Hi and thank you very much. I don't think that would work, because I need the four states of those LEDs. This box has 4 of them, almost all the time there is 1 on and the others off. But if there is an alarm, some of them light up or blink, and if there are internal problems some of them will blink slow. It is a safety-flap in the waste-water-tube, but this is not important.
Thank's anyway

**Reply by gemu2015** [reply]
anyhow you should use the optional script editor. this allows for indents and comments and color syntax . attached is a formatted output of your script which shows a bunch of unmatched if, then,else errors.

there is no syntax check in scripts (to keep firmware small) so you will not get error messages but undefined behaviour.

 
[color_script.pdf](https://github.com/arendst/Tasmota/files/8853871/color_script.pdf)

**Reply by gemu2015** [reply]
here is a suggestion for pin on measurement on an ESP32
by creating a separate task with 10 ms you get more accuracy and will not be affected by other tasks (>F may be delayed)

\>D
fcnt=0
pstate=0
ontim=0

\>B

; set pin 39 to input
spinm(39 I)

;create 10ms task
ct(1 10 1 3)


; task is called every 10 ms
; keep this part as short as possible and do not print
\>t1
; check pin state
pstate=pin[39]
if chg[pstate]>0 {
	if pstate==0 {
		;slope from 1 to 0
		ontim=fcnt
		fcnt=0
	} else {
		; slope from 0 to 1
	}
}

if pstate>0 {
	; count on time
	fcnt+=1
}

\>S

; in 10 ms steps
print ontime = %ontim%

**Comment by git5any** [comment, marked as answer]
Hi,
thank you for the help so far. I changed my code from if-then-else-endif to the {}bracket version. I'm more familiar with that and so I found the errors. The code itself was working then.
I did some more testing and now I have a solution with just one timer-variable and it works perfectly for me.
@gemu2015 Thank you for the code for pin-measurement. That looks interestiung for time critical things, and I keep that in mind. However in my case those LEDs blink either in 1sec oder 2sec frequency and in my code I already check with quite some tolerance (0.5 sec is checked between 0.3 and 0.7 sec, 1sec is checed between 0.8 and 1.2 secs). So that is tolerance enough. And again: I want those 4 LEDs, lets say, remote transferred to my homeautomation to do something with this information, if one or more LEDs change. Nothing timecritical.

```
>D
pnP=0 ;Value of Pin PowerLED
rskP=4 ;Rueckstauklappe PowerLED
oP=4 ;oldValue_RueckstauklappePowerLED
tmP=0 ;timer
ms=0 ;MQTTstate
mst="m_wait"  ;MQTTstateText
ws=0
wst="w_wait"
myTime="xx"
>B
->sensor53 r
->ButtonDebounce 130
->timezone 99
spinm(39,0)
if pin[39]<0.1 {
   rskP=0
} else {
   rskP=1
}
oP=rskP
print pin39 = %pin[39]% WiFi: %wifis%
print (boot) rskp: %rskP%  oP: %oP%
>F
pnP=pin[39]
tmP+=1
if chg[pnP]>0 { ;check for change on GPIO
   print change on GPIO
   ;tmP=0
   if pnP==0 { ;(LED off)
      print Pin = %pnP%
      if tmP>25 {
         rskP=0
      } else {
         if tmP>7
         and tmP<13 {
            rskP=3
         } else {
            if thm>2
            and tmP<8 {
               rskP=2
            }
         }
      }
      tmP=0 ; reset high counter
   } else {
      if pnP>0 {;(LED on)
         print Pin = %pnP%
         if tmP>25 {
            rskP=1
         } else {
            if tmP>7
            and tmP<13 {
               rskP=3
            } else {
               if tmP>2
               and tmP<8 {
                  rskP=2
               }
            }
         }
      }
      tmP=0 ; reset low counter
   }
   if rskP!=oP {
      print new output(byHiLo) rskP: %rskP%, (old)oP: %oP%
      oP=rskP
   }
}
if tmP>500 {
   tmP=0
}
if tmP>15
and (rskP==2 or rskP==3) {
   rskP=int(pnP)
   print new output(long) rskP: %rskP%, (old)oP: %oP%
   tmP=0
}

>S
print  (S) rskP: %rskP% (tmP: %tmP%
```
(found how to get the code formatted in the forum ;-))

Maybe someone has more information about the mpt(x) feature? The doc only says:  
measure pulse time, x>=0 defines pin to use, -1 returns low pulse time,-2 return high pulse time (if defined USE_ANGLE_FUNC)

I have USE_ANGLE_FUNC enabled but I have no idea how to work with it. Maybe it could be useful to solve my task with less code/resources.

**Reply by gemu2015** [reply]
the mpt() function does measure the high and low time of an input pin via change IRQ.  However it supports only one pin and not 4.
mpt(39) would enable change IRQ on pin 39
mpt(-1) would retrieve last high  time
mpt(-2) would retrieve last low time
in milliseconds.

**Reply by git5any** [reply]
Thank you very much!
That makes things clearer. And of cours looks like its usefull for another task, e.g reading a Watermeter to calculate flow.....
